File: week_1/day-4-sliding-window/logest_substring_follow_up/longest_substring_stream.py
stream = "abcabcbb"  # INPUT

seen = {}
left = 0
max_len = 0

for i, ch in enumerate(stream):
    if ch in seen and seen[ch] >= left:
        left = seen[ch] + 1
    seen[ch] = i
    max_len = max(max_len, i - left + 1)

print(max_len)  # Output: 3

File: week_1/day-4-sliding-window/logest_substring_follow_up/longestSubstringStream.js
const stream = "abcabcbb"; // INPUT

let map = new Map();
let left = 0, maxLen = 0;

for (let i = 0; i < stream.length; i++) {
    if (map.has(stream[i]) && map.get(stream[i]) >= left) {
        left = map.get(stream[i]) + 1;
    }
    map.set(stream[i], i);
    maxLen = Math.max(maxLen, i - left + 1);
}

console.log(maxLen); // 3
